dev/ToastNotifications/ToastNotificationUtility.cpp

```diff
@@ -0,0 +1,213 @@
+
+    std::wstring wideStringClsid{ clsid.get() };
+    RegisterValue(hKey, L"CustomActivator", reinterpret_cast<const BYTE*>(wideStringClsid.c_str()), REG_SZ, wideStringClsid.size() * sizeof(wchar_t));
+}
+
+void RegisterComServer(wil::unique_cotaskmem_string const& processName, wil::unique_cotaskmem_string const& clsid)
+{
+    wil::unique_hkey hKey;
+    //subKey: Software\Classes\CLSID\{comActivatorGuidString}\LocalServer32
+    std::wstring subKey{ c_clsIdPath + clsid.get() + LR"(\LocalServer32)" };
+
+    THROW_IF_FAILED(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr,
+        &hKey,
+        nullptr));
+
+    std::wstring comRegistrationExeString{ c_quote + processName.get() + c_quote + c_toastActivatedArgument };
+
+    RegisterValue(hKey, nullptr, reinterpret_cast<const BYTE*>(comRegistrationExeString.c_str()), REG_SZ, (comRegistrationExeString.size() * sizeof(wchar_t)));
+}
+
+void UnRegisterComServer(std::wstring const& clsid)
+{
+    wil::unique_hkey hKey;
+    //subKey: Software\Classes\CLSID\{comActivatorGuidString}\LocalServer32
+    std::wstring subKey{ c_clsIdPath + clsid + LR"(\LocalServer32)" };
+
+    THROW_IF_FAILED(RegDeleteKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        KEY_WOW64_64KEY,
+        0));
+}
+
+void UnRegisterAppIdentifierFromRegistry()
+{
+    std::wstring appIdentifier{ RetrieveAppId() };
+
+    wil::unique_hkey hKey;
+    //subKey: \Software\Classes\AppUserModelId\{AppGUID}
+    std::wstring subKey{ c_appIdentifierPath + appIdentifier };
+
+    THROW_IF_FAILED(RegDeleteKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        KEY_WOW64_64KEY,
+        0));
+}
+
+std::wstring RegisterComActivatorGuidAndAssets(winrt::Microsoft::Windows::ToastNotifications::ToastActivationInfo const& details)
+{
+    std::wstring appIdentifier{ RetrieveAppId() };
+    THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(appIdentifier.c_str(), GUID_NULL));
+
+    wil::unique_hkey hKey;
+    // subKey: \Software\Classes\AppUserModelId\{AppGUID}
+    std::wstring subKey{ c_appIdentifierPath + appIdentifier };
+
+    THROW_IF_FAILED(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr,
+        &hKey,
+        nullptr));
+
+    WCHAR registeredGuidBuffer[GUID_LENGTH + 4]; // GUID length + '{' + '}' + '\0'
+    DWORD bufferLength = sizeof(registeredGuidBuffer);
+    HRESULT status = RegGetValueW(
+        hKey.get(),
+        nullptr,
+        L"CustomActivator",
+        RRF_RT_REG_SZ,
+        nullptr,
+        &registeredGuidBuffer,
+        &bufferLength);
+
+    if (status == ERROR_FILE_NOT_FOUND)
+    {
+        wil::unique_cotaskmem_string processName;
```

Comment:
This if block needs a check: THROW_HR_IF_NULL(E_UNEXPECTED, details)

Reply:
THROW_HR_IF(E_INVALIDARG, (details == nullptr)); -  we are already checking the details in the RegisterActivator API i.e. the caller of this API.